include/ringmesh/mesh/cartesian_grid.h

```diff
@@ -0,0 +1,526 @@
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL ASGA BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *     http://www.ring-team.org
+ *
+ *     RING Project
+ *     Ecole Nationale Superieure de Geologie - GeoRessources
+ *     2 Rue du Doyen Marcel Roubault - TSA 70605
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#pragma once
+
+#include <ringmesh/basic/common.h>
+#include <ringmesh/basic/frame.h>
+#include <ringmesh/basic/geometry.h>
+#include <ringmesh/basic/logger.h>
+
+#include <ringmesh/mesh/common.h>
+
+#include <geogram/basic/attributes.h>
+
+namespace GEO
+{
+    class AttributesManager;
+} // namespace GEO
+namespace RINGMesh
+{
+    FORWARD_DECLARATION_DIMENSION_CLASS( CartesianGridBaseBuilder );
+    FORWARD_DECLARATION_DIMENSION_CLASS( CartesianGridBuilder );
+}
+
+/*!
+ * @file Generic class for creating cartesian grids
+ * @author Melchior Schuh-Senlis
+ */
+
+namespace RINGMesh
+{
+    /**
+     * Template class for Cartesian grids of different dimensions
+     * Each value of the grid represents the cell for which the point
+     * is the corner with the smallest coordinates.
+     */
+    template < index_t DIMENSION >
+    class CartesianGridBase
+    {
+        ringmesh_disable_copy_and_move( CartesianGridBase );
+        friend class CartesianGridBaseBuilder< DIMENSION >;
+
+    public:
+        /*!
+         * Constructor for the cartesian grid
+         * \param[in] nb_cells_in_each_direction number of cells in each
+         * direction of the grid.
+         * \param[in] vec_cartesian_axis used to define
+         * the cartesian grid cells : the origin of the frame is the position of
+         * the cell (0,0,0) in the cartesian grid, the vectors are the
+         * directions and length of the grid cells.
+         */
+        CartesianGridBase( ivecn< DIMENSION > nb_cells_in_each_direction,
+            ReferenceFrame< DIMENSION > vec_cartesian_axis )
+        {
+            check_and_update_number_of_cells( nb_cells_in_each_direction );
+            check_and_update_frame( vec_cartesian_axis );
+            inverse_cartesian_frame_ = ReferenceFrameManipulator< DIMENSION >::
+                orthogonal_reference_frame_from_global_to_local(
+                    cartesian_frame_ );
+            attributes_manager_.resize( nb_total_cells_ );
+        }
+
+        void save_mesh( const std::string& filename ) const
+        {
+            // TODO
```

Comment:
Throw an exception to say that is not implemented yet

Reply:
done